### ISSUE TITLE
install order should follow the order in .tool-versions

### ISSUE THREAD

**Issue by joshlartz** [asdf-vm/asdf#929]
**Describe the bug**

Some plugins depend on others to be installed first. Despite being expressed that way in the .tool-versions file, asdf alphabetizes the install regardless. 

This feels like it should be the default behavior, but at the very least there should be a flag for it. 

**To Reproduce**

```shell
# .tool-versions
python 3.9.4
awscli 2.2.0
```
`asdf install`

**Expected behavior**

The .tool-versions file is installed top to bottom.

**Actual behavior**

The installation happens in alphabetical order.

> asdf install
> * Downloading awscli release 2.2.0...
> ~/.asdf/installs/awscli/2.2.0 ~
> .asdf/plugins/awscli/bin/../lib/utils.bash: line 86: python: command not found
> .asdf/plugins/awscli/bin/../lib/utils.bash: line 87: ./venv/bin/activate: No such file or directory
> .asdf/plugins/awscli/bin/../lib/utils.bash: line 88: pip: command not found
> .asdf/plugins/awscli/bin/../lib/utils.bash: line 89: pip: command not found
> .asdf/plugins/awscli/bin/../lib/utils.bash: line 90: pip: command not found
> .asdf/plugins/awscli/bin/../lib/utils.bash: line 91: deactivate: command not found
> ~
> awscli 2.2.0 installation was successful!
> Downloading python-build...

### Environment

```shell
OS:
Darwin C02WQ018HTDD 20.4.0 Darwin Kernel Version 20.4.0: Fri Mar  5 01:14:14 PST 2021; root:xnu-7195.101.1~3/RELEASE_X86_64 x86_64

SHELL:
zsh 5.8 (x86_64-apple-darwin20.0)

ASDF VERSION:
v0.8.0
```

**Comment by jthegedus** [comment]
> @jthegedus not sure if this should be another issue, but in the regards of awscli plugin, it uses a global version of python instead of a local version, is there a way to use a locally defined tool inside a plugin?

It depends on the plugin implementation. Ideally, each tool installed is a standalone, pre-packaged binary. If it's not, the plugin may build the tool from source (tools with git-ref installation support likely do this).

As I understand, AWS CLI v2+ ships it's own version of python in pre-built packages. So either the plugin needs to handle installation of v2+ different to v1 and no longer rely on the global.

As for

>is there a way to use a locally defined tool inside a plugin?

What version is the "local" version and who is defining it? Is this defined by the plugin, by the tool being installed or by you the installer of the tool? What is it local to?

**Comment by jthegedus** [comment]
To clarify, the reason I marked this as a bug is because elsewhere in `asdf` we resolve items in `.tool-versions` line by line, top to bottom. We should be consistent and alphabetical resolution goes against that.

The solve for this is to ensure `asdf install` resolves `.tool-versions` line by line

**Comment by lifeiscontent** [comment]
@jthegedus Makes sense. At the time of writing I was thinking local to the folder you execute in. E.g. I'm in project-a and I've defined a python version in .tool-versions so I'd expect that one to be used if no python executable is found. Resolution could be recursive. Shell -> Local -> Global -> System

**Comment by jthegedus** [comment]
>Resolution could be recursive. Shell -> Local -> Global -> System

The resolution order is (from memory) Shell > `.tool-versions` in current dir and each dir up to the global `.tool-versions` file. `system` can be set in `.tool-versions` for a tool to fallback to the System installed version, or if no tool is found in `.tool-versions`, it fallsback to System.

@lifeiscontent this is probably worth a GitHub Discussion topic.

**Comment by lifeiscontent** [comment]
@jthegedus thats awesome, so if that is the case, that is not what is happening when installing awscli ~ Major version 1

**Comment by lifeiscontent** [comment]
specifically due to this line: https://github.com/MetricMike/asdf-awscli/blob/2c4638b214e2081a0f3f71aa17542981f7b6b66e/lib/utils.bash#L102

**Comment by jthegedus** [comment]
@lifeiscontent you should look at this interesting plugin someone has made - https://github.com/asdf-vm/asdf/discussions/980

**Comment by tony** [comment]
> @lifeiscontent you should look at this interesting plugin someone has made - https://github.com/asdf-vm/asdf/discussions/980

Any idea why this is a 404?

On area where this got me: [asdf-poetry](https://github.com/asdf-community/asdf-poetry) installing before [asdf-python](https://github.com/danhper/asdf-python)

**Comment by jthegedus** [comment]
>Any idea why this is a 404?

It's a 404 because [we turned off GitHub Discussions](https://github.com/asdf-vm/asdf/issues/766#issuecomment-1039154909)

>On area where this got me: [asdf-poetry](https://github.com/asdf-community/asdf-poetry) installing before [asdf-python](https://github.com/danhper/asdf-python)
 
 That is unfortunate. Sorry about this. I have marked this as a priority to be fixed, though, as always, PRs are welcome!

**Comment by 1oglop1** [comment]
I also build a plugin for another package manager PDM https://github.com/1oglop1/asdf-pdm and the installation is the same as poetry -> pipe `.py` into `python` and the same as others the installation order breaks it.

**Comment by jthegedus** [comment]
Related #1127

**Comment by dchenbecker** [comment]
Just to add another data point: the `markdownlint-cli2` plugin requires a `nodejs` install. If you haven't installed `nodejs`, the command `asdf install markdownlint-cli32 latest` happily runs and does nothing.

**Comment by chrisjpalmer** [comment]
Made a PR here to address this. https://github.com/asdf-vm/asdf/pull/1723 If interested, let me know.

**Comment by badele** [comment]
> Just to add another data point: the `markdownlint-cli2` plugin requires a `nodejs` install. If you haven't installed `nodejs`, the command `asdf install markdownlint-cli32 latest` happily runs and does nothing.

Thanks @dchenbecker  for you report, i have same issue

**Comment by stanhu** [comment]
Yes, we're having to work around this limitation for a number of dependencies:

* `rust` needs to be installed before `ruby` for YJIT support
* `nodejs` needs to be installed before `markdownlint-cli2`

I guess somebody needs to resurrect https://github.com/asdf-vm/asdf/pull/1723 and resolve the issues?

**Comment by Stratus3D** [comment]
> Some plugins depend on others to be installed first. Despite being expressed that way in the .tool-versions file, asdf alphabetizes the install regardless.

> This feels like it should be the default behavior, but at the very least there should be a flag for it.

This makes sense. You want a way of specifying the dependencies between plugins. The most direct way of doing this would be to specify plugins in the correct order so that a tool's dependencies are installed before it is. For example if tool `B` depends on tool `A` you'd specify the versions of both in the `.tool-versions` file like this:

```
A 1.2.3
B 1.4.0
```

The problem with this is that it means user's of asdf must be conscious of all inter-tool dependencies and specify the tools in the correct order in every `.tool-versions` file. Additionally, asdf would have to be able to preserve order correctly when dealing different tools defined in different `.tool-versions` files on the file system. For example, suppose a version of tool `A` was defined in a parent directory and a version of tool `B`  was defined in the current directory. Which should be installed first? We'd have to define rules for install order in asdf, and users would have to understand these rules when specifying versions. The cognitive load on users could be significant.

**Here is an alternative...**

Rather than leaving this up to users we can have plugins declare their dependencies explicitly. For example, if `B` depends on tool `A` then `B` would define a `bin/deps` callback that would print `A`. asdf would then know that the current version of `A` must always be installed before the version of `B` is installed. It wouldn't matter where the versions of `A` and `B` are defined. If `B` says it depends on `A` the appropriate version of `A` would always be installed first.

Thinking about this alternative in more detail...

**Would having a permanent dependency be limiting?**

If `B` always depends on `A` you'd always have to have `A` installed first. I don't think this would be a problem in practice. I frequently use the Erlang and Elixir plugins, and Erlang is always a dependency of Elixir. There would be no reason for a user to not need Erlang to exist first. Another example that comes to mind is the Rebar plugin. It requires Erlang in order to compile. Nothing would ever change that, the dependency (Erlang) must always be present, otherwise Rebar install/runtime will fail.

**What if a tool dependency is missing?**

Suppose a user had the Elixir plugin installed but not the Erlang plugin. There are a couple ways we could handle this:

* Fail since the Erlang plugin isn't installed. This would certainly be a breaking change for some users. 
* Log a warning and proceed without Erlang, and hope the user has another version of Erlang already installed that Elixir will use to build.

**asdf plugins don't currently specify any dependencies today. How would this feature work for plugins without a `bin/deps` callback?**

Initially we could make `bin/deps` optional. If it's absent asdf could assume the plugin does not depend on any other plugin, and versions could be installed at any time.

**How would the `bin/deps` callback work with plugin updates?**

If a plugin updated `bin/deps` and changed the declared dependencies handling things correctly could be tricky. We'd want to support plugins updating their dependencies, but a change in dependencies could require a re-install of all the versions of the tool. Having to re-install dozens of versions would be a bad user experience, but would be necessary to install versions are installed with the necessary dependencies. I'd need to give this more thought before implementing.

**Would this affect performance?**

We'd have to know more than just the current tool version when resolving shims, so performance would be impacted by this. I would not expect more than a 10-20% performance penalty from this feature for shim version resolution, but testing would be needed.

**How would this be implemented?**

I've not tried to build of proof of concept of this yet, but I think this would be challenging to implement. I think I could do it, but the changes would touch plugin updates and installs, version installs, shim generation, and version resolution. 

Implementation would require this changes:

* **Plugin updates and installs** would be updated to invoke `bin/deps`. The output would be stored somewhere in a graph that asdf would build of all inter-plugin dependencies. This would be used for inform installs and version resolution.
* **Version installs** would consult the dependency graph to determine what other tool versions would need to be installed first. Versions would be installed in the order according to the dependency graph. Installed versions would be installed into directories with names including the dependency tool and versions along with the tool versions. For example, if I installed Elixir 1.17.3 it'd end up in a directory named `1.17.3-erlang:27.1` or something like that.
* **shim generation** would be updated to handle the new install directory names.
* **version resolution** would take into account dependencies as well. For example, if I said I needed Elixir `1.17.3` with Erlang 26 but only `1.17.3-erlang:27.1` was installed version resolution would fail due to missing Elixir install with appropriate Erlang version.

Thoughts?
